Create sector `evms` for crosschain views of EVM raw tables

hildobby opened the pull request:
Brief comments on the purpose of your changes:

**For Dune Engine V2**

I've checked that:

### General checks:
* [ ] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [ ] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [ ] if adding a new model, I added a test
* [ ] the filename is unique and ends with .sql
* [ ] each sql file is a select statement and has only one view, table or function defined  
* [ ] column names are `lowercase_snake_cased`
* [ ] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [ ] if wanting to expose a model in the UI (Dune data explorer), I added a post-hook in the JINJA config to add metadata (blockchains, sector/project, name and contributor Dune usernames)

### Pricing checks:
* [ ] `coin_id` represents the ID of the coin on coinpaprika.com
* [ ] all the coins are active on coinpaprika.com (please remove inactive ones)

### Join logic:
* [ ] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

### Incremental logic:
* [ ] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * [ ] where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

hildobby commented:
Anyone knows why the `gnosis.blocks` base table has different data types? For all other EVM chains' `{chain}.blocks` tables, `difficulty` is a `BIGINT` and `total_difficulty` is a `DECIMAL(38, 0)`. But for `gnosis.blocks` they both are `UINT256` 🤷‍♂️

hildobby commented:
So only UINT256 would work here, right? What do you think is the best way forward for this to work @dot2dotseurat? Other than waiting for migration to DuneSQL to be complete I mean 😅

![Screenshot 2023-06-01 at 12 36 23](https://github.com/duneanalytics/spellbook/assets/23564601/6098d4cb-4cc9-41a9-ae85-13b7b866372a)

couralex6 commented:
> So only UINT256 would work here, right? What do you think is the best way forward for this to work @dot2dotseurat? Other than waiting for migration to DuneSQL to be complete I mean 😅
> 
> ![Screenshot 2023-06-01 at 12 36 23](https://user-images.githubusercontent.com/23564601/242570367-6098d4cb-4cc9-41a9-ae85-13b7b866372a.png)

Hey @hildobby! 

Yup, `UINT256` will eventually work here once the migration is complete and I'll make sure your spells are correctly translated.

For now, you should be able to cast to `DOUBLE` instead of `DECIMAL`. I just tried and it seems to be working. 

Lmk if that works!

jeff-dude commented on models/evms/evms_blocks.sql:
```diff
@@ -0,0 +1,143 @@
+{{ config(
+        alias ='blocks',
+        unique_key=['blockchain', 'number'],
+        post_hook='{{ expose_spells_hide_trino(\'["ethereum", "polygon", "bnb", "avalanche_c", "gnosis", "fantom", "optimism", "arbitrum"]\',
```
is this intentionally hidden?

jeff-dude replied:
same comment throughout all new spells here

hildobby replied:
ah no good catch, will change to unhide them

jeff-dude commented on models/evms/evms_blocks.sql:
```diff
@@ -0,0 +1,143 @@
+        )
+}}
+
+SELECT 'ethereum' AS blockchain
```
any reason we can't use the for loop approach, for cleaner code?

jeff-dude replied:
if it's due to `blockchain` column, there is a way to do a dictionary variable and assign a value for both blockchain and source table in same variable. rob has some examples out there, likely in nft restructure models

jeff-dude replied:
same idea across all spells here

jeff-dude commented on models/evms/evms_blocks.sql:
```diff
@@ -0,0 +1,143 @@
+{{ config(
+        alias ='blocks',
+        unique_key=['blockchain', 'number'],
```
if assigning unique keys here, do you want to apply the tests in schema file? for views, it's not vital, but this particular config setting won't do anything as-is (unless its more of a documentation kinda thing)

jeff-dude replied:
again, same applies throughout all spells here

jeff-dude reviewed (changes requested):
in addition to comments throughout, are we sure these are performant as views? if you query the cross-chain spells here, then add date filters, does it actually help or will it read entire view and then filter?

jeff-dude commented:
> I unhid the tables from trino, made them incremental materialised tables instead of views, figured out the blockchain field and changed to a loop approach and added tests @jeff-dude

what is the use case for these cross-chain base table spells? these timeouts are even more concerning, with much longer range applied too. this is a lot for spellbook to handle. it may need to be considered more at our base ingest level, or at least segmented out even further into much smaller chunks -- which doesn't seem to apply here, as it's just combining base tables.

jeff-dude commented:
i mentioned this to the ingest team. here's the option:
- revert back to views, then we can have them live in spellbook
- if the views are inefficient to query downstream, we'll have to find another process outside of spellbook to do this for you. it's not computationally efficient to do so here

hildobby commented:
> > I unhid the tables from trino, made them incremental materialised tables instead of views, figured out the blockchain field and changed to a loop approach and added tests @jeff-dude
> 
> what is the use case for these cross-chain base table spells? these timeouts are even more concerning, with much longer range applied too. this is a lot for spellbook to handle. it may need to be considered more at our base ingest level, or at least segmented out even further into much smaller chunks -- which doesn't seem to apply here, as it's just combining base tables.

Just any crosschain search really, I would use it a lot and I can imagine others too. Stuff like:
- searching crosschain balance of a wallet
- joining on other crosschain tables (even useful for some spellbook crosschain spells imo)
I have a few some wip queries that im waiting for this on since it would really reduce the number of CTEs I need

I would almost use them more than base tables, so imo it would be awesome to have it materialised somehow 🤷‍♂️

jeff-dude commented:
> > > I unhid the tables from trino, made them incremental materialised tables instead of views, figured out the blockchain field and changed to a loop approach and added tests @jeff-dude
> > 
> > 
> > what is the use case for these cross-chain base table spells? these timeouts are even more concerning, with much longer range applied too. this is a lot for spellbook to handle. it may need to be considered more at our base ingest level, or at least segmented out even further into much smaller chunks -- which doesn't seem to apply here, as it's just combining base tables.
> 
> Just any crosschain search really, I would use it a lot and I can imagine others too. Stuff like:
> 
> * searching crosschain balance of a wallet
> * joining on other crosschain tables (even useful for some spellbook crosschain spells imo)
>   I have a few some wip queries that im waiting for this on since it would really reduce the number of CTEs I need
> 
> I would almost use them more than base tables, so imo it would be awesome to have it materialised somehow 🤷‍♂️

cc @aalan3 @dsalv @nileslawrence @0xBoxer @andrewhong5297

hildobby commented:
Turned it back into views & removed tests @jeff-dude

0xBoxer commented:
I think we'll have a case where these tables are useful sooner or later so we should probably find a solution on how to construct them.

If we continue with this, I think we should use ``chainID`` as an additional identifier to make it easier and more efficient to join these tables on something else. (string vs int, also makes the naming a bit more reliable and in line with how everything else deals with different chains on a infra level)

jeff-dude reviewed (approved):
sounds good, let's give views a shot. we can revisit other comments in another forum, if needed in the future

MSilb7 commented:
> 

I think tables like this are 100000% useful (and could even be the default versus a bespoke schema per chain). I'd expect ~50% of my queries now are cross-chain & unioning over and over is annoying.

jeff-dude commented:
> > 
> 
> I think tables like this are 100000% useful (and could even be the default versus a bespoke schema per chain). I'd expect ~50% of my queries now are cross-chain & unioning over and over is annoying.

i've initiated an internal dialogue about these and potential options
